### ISSUE TITLE
Rename `unsafe_ignore_trace` and `unsafe_empty_trace`

### ISSUE THREAD

**Issue by mystor** [Manishearth/rust-gc#52]
Omitting fields from all of the trace methods is actually safe, due to the design of this garbage collector. It just means that any `Gc<T>`s which are hidden by these impls will not ever be unrooted, and thus will not be collected if they participate in a cycle.

As leaking is safe, these should be renamed from `unsafe_ignore_trace` to just `ignore_trace` and `unsafe_empty_trace` to just `empty_trace`.

**Comment by andersk** [comment]
One can cause the collector to panic with inappropriate use of `#[unsafe_ignore_trace]`. Although panicking is safe, it’s probably not what the developer expects. So perhaps it’s for the best that the name inspires some level of caution.

```rust
use gc::{Finalize, Gc, Trace};

#[derive(Finalize, Trace)]
struct S(#[unsafe_ignore_trace] Gc<()>);

fn main() {
    Gc::new(S(Gc::new(())));
}
```

**Comment by Ekleog** [comment]
I'm trying to understand what the panic message from https://github.com/Manishearth/rust-gc/issues/52#issuecomment-831672736 is about. It seems to me like the two `Gc` here should simply be independent?

In other words, I'm thinking that having `#[unsafe_ignore_trace]` should raise issues only when the field with this tag actually contains references that cause a circular dependency with the outside Gc. Because so long as the inner and outer Gc live in different worlds, I don't see how that could raise problems.

I guess, based on the error message the issue somehow stems from the fact that Gc needs to Finalize and not Drop? But I must say I'm not sure I understand what exactly Finalize is used for, so I don't get the issue.

In other words, I currently think this panic is a bug in `Gc`. But is it?

**Comment by andersk** [comment]
@Ekleog The inner `Gc` is created in the rooted state. It would ordinarily be unrooted upon being moved into the outer `Gc`; however, that’s skipped by `#[unsafe_ignore_trace]`. So the inner `Gc` is still rooted until the outer `Gc` is collected, at which time the destructor of the inner `Gc` unroots itself, triggering the finalizer of its contents. It’s generally not safe to run a finalizer at this stage during collection, because a custom finalizer might access dangling `Gc` pointers if we’re in the middle of collecting a cycle. The assertion prevents this unsafety and panics instead.

In this particular case where the finalizer is triggered from the destructor of a rooted `Gc`, there may be an argument that, even if we’re in the middle of collection, every `Gc` the finalizer might be able to reference must have been marked live, so it should be safe to skip the assertion and run the finalizer anyway. I’m not entirely confident that this argument is correct (especially in the presence of a known bug #124), but if it is, perhaps we could indeed arrange for `ignore_trace` to have no adverse consequences beyond potential leaks.

**Comment by Ekleog** [comment]
Hmm… do you know of a place where I could read about what `Finalize` is supposed to do? I guess I need to read some more than the one-line doc-comment of `Finalize` and `Trace` in order to understand why this assertion is even necessary for finalizing unrooted pointers :)
